spec/index.md

```diff
@@ -667,35 +667,55 @@ bar: map<address, u256> # bar is assigned a static nonce by the compiler
 baz: map<address, map<address, u256>> # baz is assigned a static nonce by the compiler
 ```
 
-The expression `bar[0x00]` would resolve to the hash of both bar's nonce and the key value 
-.i.e. `keccak256(<bar nonce>, 0x00)`. Similarly, the expression `baz[0x00][0x01]` would resolve to 
+The expression `bar[0x00]` would resolve to the hash of both bar's nonce and the key value
+.i.e. `keccak256(<bar nonce>, 0x00)`. Similarly, the expression `baz[0x00][0x01]` would resolve to
 a nested hash i.e. `keccak256(keccak256(<baz nonce>, 0x00), 0x01)`.
 
+#### 6.2.3. The `to_mem` function
+
+Reference type values can be copied to memory using the `to_mem` function.
+
+Example:
+
+```python
+my_array_var: u256[10] = self.my_array_field.to_mem()
+```
+
 ### 6.3. Memory
 
 Only sequence types can be stored in memory.
 
-The first memory slot (`0x00`) is used to keep track of the lowest available memory slot. Newly 
+The first memory slot (`0x00`) is used to keep track of the lowest available memory slot. Newly
 allocated segments begin at the value given by this slot. When more memory has been allocated,
 the value stored in `0x00` is increased.
 
 We do not free memory after it is allocated.
 
 #### 6.3.1. Sequence types in memory
 
-Sequence type values may exceed the 256-bit stack slot size, so we store them in memory and 
+Sequence type values may exceed the 256-bit stack slot size, so we store them in memory and
 reference them using pointers kept on the stack.
 
 Example:
 
 ```python
 # function scope
-foo: u256[100] # foo is a pointer that references 100 * 256 bits in memory. 
+foo: u256[100] # foo is a pointer that references 100 * 256 bits in memory.
 ```
 
 To find an element inside of a sequence type, the relative location of the element is added to the
 given pointer.
 
+#### 6.2.3. The `clone` function
+
+Reference type values can be cloned using the `clone` function.
+
+Example:
+
+```python
+foo: u256[10] = bar.clone()
```

Comment:
Maybe this could be expanded to use some `assert`s to make the behavior more clear.

```
foo: u256[10] = bar.clone()
assert foo[1] == bar[1]
foo[1] = 1
assert foo[1] != bar[1]
```

We could even add in another section that explains what would happen if we don't use `clone`  e.g.

```
foo: u256[10] = bar
assert foo[1] == bar[1]
foo[1] = 1
assert foo[1] == bar[1]
```

Reply:
As per the example, it seems the cloning only copies the value instead of the pointer reference?

Reply:
Correct. Use `clone` to get an independent copy, omit clone and you actually copy the reference instead. We have tests that demonstrate it but I think it would make sense to expand the example right here.

Reply:
Yeah, makes sense @cburgdorf.